### PR TITLE
[SPARK-36327][SQL] Spark sql creates staging dir inside database directory rather than creating inside table directory

### PR DESCRIPTION
**Pull request by senthh**
### What changes were proposed in this pull request?

This PR does minor changes in the file SaveAsHiveFile.scala.

It contains the below changes :

1. dropping getParent from below part of code
===============================
if (extURI.getScheme == "viewfs") {
getExtTmpPathRelTo(path.getParent, hadoopConf, stagingDir)
===============================

### Why are the changes needed?

Hive is creating .staging directories inside "/db/table" location but Spark-sql creates .staging directories inside /db/" location when we use hadoop federation(viewFs). But works as expected (creating .staging inside /db/table/ location for other filesystems like hdfs).

In HIVE:
```
 beeline
> use dicedb;
> insert into table part_test partition (j=1) values (1);
...
INFO  : Loading data to table dicedb.part_test partition (j=1) from **viewfs://cloudera/user/daisuke/dicedb/part_test/j=1/.hive-staging_hive_2021-07-19_13-04-44_989_6775328876605030677-1/-ext-10000** 
```

but spark's behaviour,

```
spark-sql> use dicedb;
spark-sql> insert into table part_test partition (j=2) values (2);
21/07/19 13:07:37 INFO FileUtils: Creating directory if it doesn't exist: **viewfs://cloudera/user/daisuke/dicedb/.hive-staging_hive_2021-07-19_13-07-37_317_5083528872437596950-1**
... 
```

The reason why we require this change is , if we allow spark-sql to create .staging directory inside /db/ location then we will end-up with security issues. We need to provide permission for "viewfs:///db/" location to all users who submit spark jobs.

After this change is applied spark-sql creates .staging inside /db/table/,  similar to hive, as below,

```
spark-sql> use dicedb;
21/07/28 00:22:47 INFO SparkSQLCLIDriver: Time taken: 0.929 seconds
spark-sql> insert into table part_test partition (j=8) values (8);
21/07/28 00:23:25 INFO HiveMetaStoreClient: Closed a connection to metastore, current connections: 1
21/07/28 00:23:26 INFO FileUtils: Creating directory if it doesn't exist: **viewfs://cloudera/user/daisuke/dicedb/part_test/.hive-staging_hive_2021-07-28_00-23-26_109_4548714524589026450-1** 
```

The reason why we don't see this issue in Hive but only occurs in Spark-sql:

In hive, "/db/table/tmp" directory structure is passed for path and hence path.getParent returns "db/table/" . But in Spark we just pass "/db/table" so it is not required to use "path.getParent" for hadoop federation(viewfs)

### Does this PR introduce _any_ user-facing change?
 No

### How was this patch tested?

Tested manually by creating hive-sql.jar

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala [hunk 1]
```diff
@@ -188,7 +188,7 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
       stagingDir: String): Path = {
     val extURI: URI = path.toUri
     if (extURI.getScheme == "viewfs") {
-      getExtTmpPathRelTo(path.getParent, hadoopConf, stagingDir)
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala:191]: It says it was copied from `Context.java#getExternalTmpPath` in Hive 1.2. Can you check how it's fixed in Hive?

Comment (senthh) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala:191]: In hive, "/db/table/tmp" directory structure is passed for path and hence path.getParent returns "db/table/" . But in Spark we just pass "/db/table" so it is not required to use "path.getParent" for hadoop federation(viewfs)

**Review by dongjoon-hyun — COMMENTED** [review]
So, is this only for `viewfs`, @senthh ?

**Comment by dongjoon-hyun** [comment]
cc @sunchao

**Comment by senthh** [comment]
> So, is this only for `viewfs`, @senthh ?

yes, we see this issue only for viewfs. And works fine if it is "hdfs://"

**Comment by sunchao** [comment]
@senthh could you explain the motivation in this PR in the "Why are the changes needed?" section? IIRC Hive also stores staging directory under database directory so I'm wondering why we need the change here.

**Comment by senthh** [comment]
> @senthh could you explain the motivation in this PR in the "Why are the changes needed?" section? IIRC Hive also stores staging directory under database directory so I'm wondering why we need the change here.

@sunchao Hive is creating .staging directories inside "/db/table" location but Spark-sql creates .staging directories inside /db/" location when we use hadoop federation(viewFs). But works as expected (creating .staging inside /db/table/ location for other filesystems like hdfs).

HIVE:
{{
# beeline
> use dicedb;
> insert into table part_test partition (j=1) values (1);
...
INFO  : Loading data to table dicedb.part_test partition (j=1) from **viewfs://cloudera/user/daisuke/dicedb/part_test/j=1/.hive-staging_hive_2021-07-19_13-04-44_989_6775328876605030677-1/-ext-10000** 

}}

but spark's behaviour,

{{
spark-sql> use dicedb;
spark-sql> insert into table part_test partition (j=2) values (2);
21/07/19 13:07:37 INFO FileUtils: Creating directory if it doesn't exist: **viewfs://cloudera/user/daisuke/dicedb/.hive-staging_hive_2021-07-19_13-07-37_317_5083528872437596950-1**
... 
}}


The reason why we require this change is , if we allow spark-sql to create .staging directory inside /db/ location then we will end-up with security issues. We need to provide permission for "viewfs:///db/" location to all users who submit spark jobs.

After this change is applied spark-sql creates .staging inside /db/table/,  similar to hive, as below,

{{
spark-sql> use dicedb;
21/07/28 00:22:47 INFO SparkSQLCLIDriver: Time taken: 0.929 seconds
spark-sql> insert into table part_test partition (j=8) values (8);
21/07/28 00:23:25 INFO HiveMetaStoreClient: Closed a connection to metastore, current connections: 1
21/07/28 00:23:26 INFO FileUtils: Creating directory if it doesn't exist: **viewfs://cloudera/user/daisuke/dicedb/part_test/.hive-staging_hive_2021-07-28_00-23-26_109_4548714524589026450-1** 
}} 

The reason why we don't see this issue in Hive but only occurs in Spark-sql:

In hive, "/db/table/tmp" directory structure is passed for path and hence path.getParent returns "db/table/" . But in Spark we just pass "/db/table" so it is not required to use "path.getParent" for hadoop federation(viewfs)

**Comment by dongjoon-hyun** [comment]
Is there a chance that we can add a test coverage for this `viewfs` issue, @senthh ?

**Comment by senthh** [comment]
@dongjoon-hyun , we tested in our internal cluster also with Customer's cluster manually. As this issue occur only when viewfs is configured in cluster,  I think test coverage that meets viewfs will not be possible. 

But we can test this scenario manually if we have configured viewFs in our cluster. And I have added manual test results in the previous post.

**Comment by senthh** [comment]
@dongjoon-hyun Could you please let me know if anything I can do to make this PR to the further steps?


Note: I have created a jira [1]i and working n order to look opportunity to create test coverage that matches viewfs scenario.

[1] https://issues.apache.org/jira/browse/SPARK-36412

**Comment by senthh** [comment]
@HyukjinKwon @sunchao @dongjoon-hyun 

Could you be kind to let me know the next steps to make this PR into the next step?

**Review by dongjoon-hyun — APPROVED** [review]
Thank you for this and filing SPARK-36412, @senthh and all.

I revised the PR description according to the discussion.

**Review by sunchao — APPROVED** [review]
Thanks for the ping @senthh. This LGTM too. I guess similar thing can happen if people are using Router-based Federation in HDFS, but it's much harder to check the case.

**Pull request closed without merging**